packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -70,217 +73,375 @@ fn get_pkg_path(name: &str, pkg_type: PackageType) -> String {
+            })?
+            .try_into()
+            .map_err(|e| {
+                JSRunnerError::V8(format!(
+                    "Could not convert value at index 4 in runner.js as a function: {e}"
+                ))
+            })?;
+        let state_interim_sync: v8::Local<'_, v8::Function> = fns
+            .get_index(scope, 5)
+            .ok_or_else(|| {
+                JSRunnerError::V8("Could not get index 5 from runner.js array".to_string())
+            })?
+            .try_into()
+            .map_err(|e| {
+                JSRunnerError::V8(format!(
+                    "Could not convert value at index 5 in runner.js as a function: {e}"
+                ))
+            })?;
+        let state_snapshot_sync: v8::Local<'_, v8::Function> = fns
+            .get_index(scope, 6)
+            .ok_or_else(|| {
+                JSRunnerError::V8("Could not get index 6 from runner.js array".to_string())
+            })?
+            .try_into()
+            .map_err(|e| {
+                JSRunnerError::V8(format!(
+                    "Could not convert value at index 6 in runner.js as a function: {e}"
+                ))
+            })?;
+
+        Ok(Embedded {
             hash_stdlib,
             hash_util,
-            start_experiment: fns.get(0)?,
-            start_sim: fns.get(1)?,
-            run_task: fns.get(2)?,
-            ctx_batch_sync: fns.get(3)?,
-            state_sync: fns.get(4)?,
-            state_interim_sync: fns.get(5)?,
-            state_snapshot_sync: fns.get(6)?,
+            start_experiment,
+            start_sim,
+            run_task,
+            ctx_batch_sync,
+            state_sync,
+            state_interim_sync,
+            state_snapshot_sync,
         })
     }
 }
 
-/// Due to flushing, need batches and schemas in both Rust and JS.
 struct SimState {
     agent_schema: Arc<Schema>,
     msg_schema: Arc<Schema>,
 }
 
-struct RunnerImpl<'m> {
-    embedded: Embedded<'m>,
-    this: mv8::Value<'m>,
+struct ThreadLocalRunner<'s> {
+    embedded: Embedded<'s>,
+    this: v8::Local<'s, v8::Value>,
     sims_state: HashMap<SimulationShortId, SimState>,
 }
 
-// we pass in _mv8 for the return values lifetime
-fn sim_id_to_js(_mv8: &MiniV8, sim_id: SimulationShortId) -> mv8::Value<'_> {
-    mv8::Value::Number(sim_id as f64)
+fn sim_id_to_js<'s>(
+    scope: &mut v8::HandleScope<'s>,
+    sim_id: SimulationShortId,
+) -> v8::Local<'s, v8::Value> {
+    v8::Number::new(scope, sim_id as f64).into()
 }
 
-// we pass in _mv8 for the return values lifetime
-fn pkg_id_to_js(_mv8: &MiniV8, pkg_id: PackageId) -> mv8::Value<'_> {
-    mv8::Value::Number(pkg_id.as_usize() as f64)
+fn pkg_id_to_js<'s>(
+    scope: &mut v8::HandleScope<'s>,
+    pkg_id: PackageId,
+) -> v8::Local<'s, v8::Value> {
+    v8::Number::new(scope, pkg_id.as_usize() as f64).into()
 }
 
-fn idxs_to_js<'m>(mv8: &'m MiniV8, idxs: &[usize]) -> Result<mv8::Value<'m>> {
-    let a = mv8.create_array();
+fn idxs_to_js<'s>(
```

Comment:
While touching this function, could we rename this function to be more descriptive? Or add a doc-string?

Reply:
Addressed in https://github.com/hashintel/hash/pull/414/commits/a1edd4e1adc83b2e5a22f733a9254886d07ac9fb